### Problem 3### 

The prime factors of 13195 are 5, 7, 13 and 29.

What is the largest prime factor of the number 600851475143 ?

In [1]:
#first idea:
#we can just find all prime factors up until number x and then list them.

In [2]:
a = 600851475143

In [3]:
primes  = []
for i in range(1000):
    if a%(i+1) == 0:
        primes.append(i+1)

In [4]:
primes

[1, 71, 839]

In [5]:
# since primes[2]/primes[1] isn't a integer, we can now take a new_a that is a/primesp

primesp = primes[2]*primes[1]

In [6]:
new_a = int(a/primesp)

In [7]:
new_a

10086647

there are no prime factors less than 1000, but we reduced our search by a lot...

now we can check all the numbers until rootsqr ( new_a)
if we dont find any prime, then new_a is a prime and is bigger than 71  and 839.

the <b>smallest</b> prime factor that is less than or equal to $\sqrt{n}$, unless $n$ is prime.
One proof is as follows: Suppose $n=ab$ and $a$ is the smallest prime factor of $n$, and $n$ is not prime.
Since $n$ is not prime, we have $b\ne1$.  Since $a$ is the smallest prime factor of $n$, we have $a\le b$.
If $a$ were bigger than $\sqrt{n}$, then $b$ would also be bigger than $\sqrt{n}$, so $ab$ would be bigger than $\sqrt{n}\cdot\sqrt{n}$.  But $ab=n$.


now we can check all the numbers until rootsqr ( new_a)
if we dont find any prime, then new_a is a prime and is bigger than 71  and 839.

In [8]:
b = (new_a)**(1/2)

In [9]:
b = 3176

In [10]:
primes2 = []
for i in range(b):
    if new_a%(i+1) == 0:
        primes2.append(i+1)

In [11]:
primes2

[1, 1471]

In [12]:
new_a/primes2[1]

6857.0

In [13]:
new2_a = 6857

In [14]:
primes3 = []
for i in range(b):
    if new2_a%(i+1) == 0:
        primes3.append(i+1)

In [15]:
primes3

[1]

therefore new2_a is the largest prime factor!

In [45]:
def kep(x):
    return int((x**(1/2)) // 1) + 1

In [50]:
def func(x):
    pek = kep(x) # aqui encontramos uma formula meio meh pra int da sqrt
    pprimes = []  # lista vazia n1
    for i in range(2,pek): # aqui vamos pular o 1 visto q ele divide todos
        if x%i == 0:
            pprimes.append(i)
    leng = len(pprimes)
    if leng == 0: # se nao encontrarmos entao ele é primo
        return x
    else:
        for j in range(leng): # aqui vamos começar o teste de primalidade dos elementos de pprimes
            for pp in pprimes:
                if pp%pprimes[j] == 0:
                    
        

In [52]:
func(64)

[2, 4, 8]

In [58]:
func(72)

[2, 3, 4, 6, 8]

In [34]:
kep(10)

3

In [35]:
func(5)

In [36]:
kep(5)

2